### ISSUE TITLE
quarkus-keycloak-admin-client-reactive throwing BlockingNotAllowedException from ServerRequestFilter

### ISSUE THREAD

**Issue by TiagoGrosso** [quarkusio/quarkus#29786]
### Describe the bug

I'm using Quarkus with `quarkus-resteasy-reactive` and the `quarkus-keycloak-admin-client-reactive` extensions. I'm building a `ServerRequestFilter` that queries Keycloak for a user attribute, but even though I'm using the reactive client, I'm getting `BlockingNotAllowedException` when calling its methods.

The filter is this:

```java
public class Filters {
    @Inject
    Keycloak keycloak;

    @ServerRequestFilter
    public Uni<Response> filter(ContainerRequestContext requestContext) {
        return Uni.createFrom().item(() ->
                keycloak.realm("my-realm")
                        .users()
                        .search("user-that-i-get-from-context")
                        .stream()
                        .findFirst()
                        .orElseThrow()
                        .firstAttribute("the-attribute")).map(attr -> {
            if (attr.equals("some-value")){
                return null;
            }
            return Response.status(403).build();
        });
    }
}
```

In the example here I wrapped the call to Keycloak inside a `Uni` for the sake of simplicity in reporting the bug because in actuality what I want to have is a query to the DB first and then I'll call Keycloak, so the filter really must return a `Uni`.

I am able to use the Keycloak client directly in my endpoints, this only happens when using it in a `ServerRequestFilter`.

I've tried adding `.runSubscriptionOn(Infrastructure.getDefaultWorkerPool())` to the above `Uni`, as well as wrapping it in `vertx.executeBlocking()` with no success, I still get the exception.

This also only seems to happen on endpoints that return a `Uni`. On endpoints that simply return a `Response` or another type, the exception does not occur

### Expected behavior

Since I'm using the keycloak reactive client, there should be no blocking and therefore there should be no `BlockingNotAllowedException` anywhere its methods are called

### Actual behavior

I get a `BlockingNotAllowedException` when calling methods of the Keycloak reactive client in a `ServerRequestFilter`

### How to Reproduce?

You can use this project to reproduce: https://github.com/TiagoGrosso/keycloak-quarkus-reactive-bug

Simply send a request to GET http://localhost:8080/hello

If you switch out the return of the endpoint in `GreetingResource` to `String` the bug stops happening.

### Output of `uname -a` or `ver`

Ubuntu 5.15.0-56-generic #62~20.04.1-Ubuntu SMP Tue Nov 22 21:24:20 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk 17.0.2 2022-01-18

### Quarkus version or git rev

2.14.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)

### Additional information

Full stacktrace:

```java
ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-0) Request failed: org.jboss.resteasy.reactive.common.core.BlockingNotAllowedException
	at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.unwrap(InvocationBuilderImpl.java:210)
	at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.method(InvocationBuilderImpl.java:329)
	at org.keycloak.admin.client.resource.UsersResourcecec060e01a4e8719a5229b25492913298e3846fc54.search(Unknown Source)
	at com.tiagop.filters.Filters.lambda$filter$0(Filters.java:20)
	at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:28)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:52)
	at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:112)
	at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:89)
	at org.jboss.resteasy.reactive.server.filters.FilterUtil.handleUniResponse(FilterUtil.java:74)
	at com.tiagop.filters.Filters$GeneratedServerRequestFilter$filter.filter(Unknown Source)
	at org.jboss.resteasy.reactive.server.spi.ResteasyReactiveContainerRequestFilter.filter(ResteasyReactiveContainerRequestFilter.java:11)
	at org.jboss.resteasy.reactive.server.handlers.ResourceRequestFilterHandler.handle(ResourceRequestFilterHandler.java:48)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:104)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
	at org.jboss.resteasy.reactive.server.handlers.RestInitialHandler.beginProcessing(RestInitialHandler.java:48)
	at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:23)
	at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:10)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:84)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:71)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:136)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.lambda$sendStatic$1(StaticHandlerImpl.java:290)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by sberyozkin** [comment]
@TiagoGrosso I believe, according to https://quarkus.io/guides/resteasy-reactive#request-or-response-filters, you need to have `@Blocking` added to the endpoint method(s) this `ServerRequestFilter` runs before, CC @geoand

**Comment by TiagoGrosso** [comment]
> @TiagoGrosso I believe, according to https://quarkus.io/guides/resteasy-reactive#request-or-response-filters, you need to have `@Blocking` added to the endpoint method(s) this `ServerRequestFilter` runs before, CC @geoand

@sberyozkin this filter will run on most of my endpoints, so that would defeat the purpose of using resteasy-reactive if the majority of what I'll be doing is using `@Blocking`. 

The root of this problem is that I'm supposedly using a **non-blocking** client (`quarkus-keycloak-admin-client-reactive`), so I don't understand how blocking is even occurring here

**Comment by geoand** [comment]
Answered on [StackOverflow](https://stackoverflow.com/a/74767272/2504224).

The problem here is that the `Keycloack` object does not provide a Mutiny based API even though under the hood the IO is non-blocking.

**Comment by TiagoGrosso** [comment]
@geoand I'm sorry but that answer doesn't seem to cover the whole story in my eyes so allow me a few follow-up questions to better understand this on my side:

1. How come this only happens inside a `ServerRequestFilter`? Why can I use the Keycloak client directly in the endpoints without this issue?
2. If the IO is non-blocking, how is the `BlockingNotAllowedException` produced? Isn't that exception thrown only when a blocking method is called?
3. I've dug around a bit more and when the endpoint is returning a `Uni` I always get the exception, but when the endpoint is **not** returning a `Uni`, then the **first time** I call it, it throws the exception and **subsequent calls are ok and don't throw the exception**. How is this behavior not a bug? This is the actual filter code: 

```java
    @ServerRequestFilter
    public Uni<Response> filter(ContainerRequestContext requestContext) {
        final String keyValue = requestContext.getUriInfo().getQueryParameters().getFirst("apiKey");
        if (keyValue == null) {
            return Uni.createFrom().item(Response.status(403).build());
        }

        // apiKeyRepository is a simple hibernate-reactive-panache repository
        return apiKeyRepository.getApiKey(keyValue).chain(apiKey -> {
            // This eventually calls keycloak.realm("realm").users().search() where the exception is thrown
            final long limit = keycloakService.getUserLimit(apiKey.username);
            if (limit < 0) {
                return Uni.createFrom().nullItem();
            }
            // This one uses the Quarkus Redis client
            return rateLimitingService.checkAndInc(apiKey.username, limit).map(
                    canProceed -> canProceed ? null : Response.status(403).build()
            );
        });
    }
```

**Comment by geoand** [comment]
> How come this only happens inside a ServerRequestFilter? Why can I use the Keycloak client directly in the endpoints without this issue?

This happens because your endpoints are blocking while the filters run on the event loop by default.

> If the IO is non-blocking, how is the BlockingNotAllowedException produced? Isn't that exception thrown only when a blocking method is called?

A synchronous API that uses non-blocking IO under the hood needs to block the current thread in order to honor the synchronous nature of the return type.

For you last point, if you attach a sample, I will check it out

**Comment by TiagoGrosso** [comment]
@geoand thanks for your patience, I really appreciate it and you've provided me with great pointers to investigate this myself and understand the problem. It's now clearer to me. Will just leave my impressions on this comment if someone in the future has a similar question

> This happens because your endpoints are blocking while the filters run on the event loop by default.

Indeed. I was under the impression that I was using `Keycloak` directly in more places but it turns all the endpoints that call it do not return a `Uni` and are running on the executor thread, not the event loop, so I guess that means they are blocking even though they are not annotated with `@Blocking`

>A synchronous API that uses non-blocking IO under the hood needs to block the current thread in order to honor the synchronous nature of the return type.

Got it, makes sense!

> For you last point, if you attach a sample, I will check it out

I tried to create a minimum reproducible sample to send your way, but in doing so I found out the quirk. I had actually added caching to `apiKeyRepository`, which meant that on the second request, the `getApiKey` method was returning a completed `Uni`. In doing so, the code inside the `chain` method was actually running on the executor thread, because the execution did not transition to the event loop. When the value was not cached, i.e., on the first request, since it required a call to the DB, the code inside the chain was running in the event loop. Hence why when the endpoint itself returned a `Uni`, the problem happened everytime, because everything in the filter was already running inside the event loop

We can wrap this discussion here, the problem is clear to me. Thank you once again, learned a lot

**Comment by geoand** [comment]
Glad to hear it :)

**Comment by Eric-Prokop** [comment]
Sorry for commenting on a rather old issue but I want to make sure I understood correctly.

The current (2025) version of the extension [Keycloak Admin REST Client](https://quarkus.io/extensions/io.quarkus/quarkus-keycloak-admin-rest-client/) uses reactive IO underneath, but we still need use the blocking API since Keycloak iteself does not provide a Mutiny based API?

So is that what happens?
1. *Vertex Event Loop Thread*: This is the main Vertex thread that accepted the request
2. *Vertex Worker Thread*: Since I annotated my code with `@Blocking` the request is passed to a work thread right away
3. *Keycloak Reactive IO Thread*: When calling a method on the `Keycloak` instance (in the work thread), IO is handled in a reactive way, so I suppose that there is a single thread that handels all IO for every call to the `Keycloak` instance (despite from which Vertex Worker Thread the call originated)

Did I understand that correctly?

Now suppose I would clone the repo and change the return type of every `@Path` annotated method in https://github.com/keycloak/keycloak/tree/main/integration/admin-client/src/main/java/org/keycloak/admin/client/resource to return Mutiny `Uni`s and `Multi`s, could I use that then with the [Quarkus REST Client that supports async calls](https://quarkus.io/guides/rest-client#async-support) in a truly reactive way?

Also on a other note, I noticed that there was recent activity in regards to the Keycloak client (4d6af2f4a8c608e2ce8ca4795e149d145d3940d8) so might there currently already be plans to provide a Mutiny based API?
